packages/flutter/lib/src/widgets/app_lifecycle_listener.dart

```diff
@@ -0,0 +1,248 @@
+
+  /// A callback that is called when a view in the application gains input
+  /// focus.
+  ///
+  /// A call to this callback indicates that the application is entering a state
+  /// where it is visible, active, and accepting user input.
+  final VoidCallback? onResume;
+
+  /// A callback that is called when the application is hidden.
+  ///
+  /// On mobile platforms, this is usually just before the application is
+  /// replaced by another application in the foreground.
+  ///
+  /// On desktop platforms, this is just before the application is hidden by
+  /// being minimized or otherwise hiding all views of the application.
+  ///
+  /// On the web, this is just before a window (or tab) is hidden.
+  final VoidCallback? onHide;
+
+  /// A callback that is called when the application is shown.
+  ///
+  /// On mobile platforms, this is usually just before the application replaces
+  /// another application in the foreground.
+  ///
+  /// On desktop platforms, this is just before the application is shown after
+  /// being minimized or otherwise made to show at least one view of the
+  /// application.
+  ///
+  /// On the web, this is just before a window (or tab) is shown.
+  final VoidCallback? onShow;
+
+  /// A callback that is called when the application is paused.
+  ///
+  /// On mobile platforms, this happens right before the application is replaced
+  /// by another application.
+  ///
+  /// On desktop platforms and the web, this function is not called.
+  final VoidCallback? onPause;
+
+  /// A callback that is called when the application is resumed after being
+  /// paused.
+  ///
+  /// On mobile platforms, this happens just before this application takes over
+  /// as the active application.
+  ///
+  /// On desktop platforms and the web, this function is not called.
+  final VoidCallback? onRestart;
+
+  /// A callback used to ask the application if it will allow exiting the
+  /// application for cases where the exit is cancelable.
+  ///
+  /// Exiting the application isn't always cancelable, but when it is, this
+  /// function will be called before exit occurs.
+  ///
+  /// Responding [AppExitResponse.exit] will continue termination, and
+  /// responding [AppExitResponse.cancel] will cancel it. If termination is not
+  /// canceled, the application will immediately exit.
+  final AppExitRequestCallback? onExitRequested;
+
+  /// A callback that is called when an application has exited, and detached all
+  /// host views from the engine.
+  ///
+  /// This callback is only called on iOS and Android.
+  final VoidCallback? onDetach;
+
+  bool _debugDisposed = false;
+
+  /// Call when the listener is no longer in use.
+  ///
+  /// Do not use the object after calling [dispose].
+  ///
+  /// Subclasses must call this method in their overridden [dispose], if any.
+  @mustCallSuper
+  void dispose() {
+    assert(_debugAssertNotDisposed());
+    binding.removeObserver(this);
+    _debugDisposed = true;
```

Comment:
nit: set this inside an assert so the field can potentially be tree shaken out as unused in release mode?

Reply:
Done.